File: competition/agents/expander_python/agent.py
"""
Edit this file to implement your agent.

`Agent.act(obs)` is called once per turn. The `obs` argument is built by
`main.py` from the wire-protocol frame and has these fields:

    obs.H, obs.W            board dimensions (constant for the whole game)
    obs.turn                current turn number, increments each step
    obs.my_land             total cells you own
    obs.my_army             total armies summed over your cells
    obs.opp_land            opponent's land count (visible at all times)
    obs.opp_army            opponent's army total (visible at all times)
    obs.type_grid[r][c]     0=fog, 1=plain, 2=mountain, 3=city, 4=general, 5=structure-in-fog
    obs.owner_grid[r][c]    0=neutral/unknown, 1=me, 2=opp  (perspective-relative)
    obs.army_grid[r][c]     army count, 0 in fog or empty

`act` must return a 5-tuple `(pass, row, col, direction, split)`:

    pass:       1 to skip the turn, 0 to move
    row, col:   source cell (must be owned by you and have army > 1)
    direction:  0=up, 1=down, 2=left, 3=right
    split:      0=move all-but-one armies, 1=move half (floor division)

Invalid moves are silently treated as a pass by the engine.
"""

# A no-op action — used when no valid move exists or as a safe default.
PASS = (1, 0, 0, 0, 0)

# (dr, dc) offsets for direction codes 0..3
DIRECTIONS = [(-1, 0), (1, 0), (0, -1), (0, 1)]


def _is_passable(t):
    # Mountains (2) and fogged-structures (5) are impassable; everything else
    # can be entered (including fog — you just don't know what's there).
    return t != 2 and t != 5


class Agent:
    """Expander strategy.

    Each turn pick the move that maximizes
        score = src_army * (10 if expansion else 1) * (2 if opponent else 1)
    among captures (src_army > dest_army + 1). If no capture is possible
    but some legal move exists, take the first one. Otherwise pass.
    """

    def __init__(self, player_id, H, W):
        # We get the static game info once at startup. You don't have to
        # store any of it on the agent if you don't want to.
        self.player_id = player_id
        self.H = H
        self.W = W

    def act(self, obs):
        best_score = -1.0
        best_move = None
        first_valid = None

        # Scan every cell on the board. The expander only ever moves armies
        # *out* of cells it already owns, so we can skip everything else.
        for r in range(obs.H):
            for c in range(obs.W):
                if obs.owner_grid[r][c] != 1:
                    continue
                src_army = obs.army_grid[r][c]
                # Need at least 2 armies: one always stays behind on the source.
                if src_army <= 1:
                    continue

                # Try each of the four neighbor cells.
                for d, (dr, dc) in enumerate(DIRECTIONS):
                    nr, nc = r + dr, c + dc
                    if not (0 <= nr < obs.H and 0 <= nc < obs.W):
                        continue
                    if not _is_passable(obs.type_grid[nr][nc]):
                        continue

                    move = (0, r, c, d, 0)
                    # Remember any legal move as a fallback (used when no
                    # cell has enough army to actually capture anything).
                    if first_valid is None:
                        first_valid = move

                    dest_owner = obs.owner_grid[nr][nc]
                    dest_army = obs.army_grid[nr][nc]
                    # To capture, we need strictly more army than what's there
                    # (since one must stay on the source).
                    if src_army <= dest_army + 1:
                        continue

                    # Expansion = claiming new visible territory (vs reinforcing
                    # one of our own cells).
                    is_opp = dest_owner == 2
                    dest_type = obs.type_grid[nr][nc]
                    is_visible_neutral = (dest_owner == 0) and dest_type not in (0, 5)
                    is_expansion = is_opp or is_visible_neutral

                    # Bigger army = stronger move. Expansion is much more
                    # valuable than reinforcing; capturing the opponent
                    # specifically is worth double again.
                    score = float(src_army)
                    if is_expansion:
                        score *= 10.0
                    if is_opp:
                        score *= 2.0

                    if score > best_score:
                        best_score = score
                        best_move = move

        # Prefer the best capture; else any legal move; else pass.
        if best_move is not None:
            return best_move
        if first_valid is not None:
            return first_valid
        return PASS

File: competition/agents/expander_rust/src/agent.rs
// Edit this file to implement your agent.
//
// `Observation` is the parsed per-turn view from main.rs.
// `Agent::act` is called once per turn and must return an `Action`:
//
//     pass:      1 to skip the turn, 0 to move
//     row, col:  source cell
//     dir:       0=up, 1=down, 2=left, 3=right
//     split:     0=move all-but-one armies, 1=move half (floor)
//
// Tile-type codes in obs.type_grid:
//     0=fog, 1=plain, 2=mountain, 3=city, 4=general, 5=structure-in-fog
//
// Owner codes in obs.owner_grid (perspective-relative):
//     0=neutral/unknown, 1=me, 2=opp

pub struct Observation {
    pub h: usize,
    pub w: usize,
    pub turn: i32,
    pub my_land: i32,
    pub my_army: i32,
    pub opp_land: i32,
    pub opp_army: i32,
    pub type_grid: Vec<i32>,    // h*w row-major: type_grid[r * w + c]
    pub owner_grid: Vec<i32>,
    pub army_grid: Vec<i32>,
}

#[derive(Copy, Clone)]
pub struct Action {
    pub pass: i32,
    pub row: i32,
    pub col: i32,
    pub dir: i32,
    pub split: i32,
}

const PASS: Action = Action { pass: 1, row: 0, col: 0, dir: 0, split: 0 };

pub struct Agent {
    pub player_id: i32,
    pub h: usize,
    pub w: usize,
}

impl Agent {
    pub fn new(player_id: i32, h: usize, w: usize) -> Self {
        Self { player_id, h, w }
    }

    /// Expander: each turn pick the move that maximizes
    ///   score = src_army * (10 if expansion else 1) * (2 if opponent else 1)
    /// among captures (src_army > dest_army + 1). Otherwise the first valid
    /// move; otherwise pass.
    pub fn act(&self, obs: &Observation) -> Action {
        const DR: [i32; 4] = [-1, 1, 0, 0];   // up, down, left, right
        const DC: [i32; 4] = [ 0, 0, -1, 1];

        let h = obs.h as i32;
        let w = obs.w as i32;

        let mut best = PASS;
        let mut best_score: f64 = -1.0;
        let mut first_valid: Option<Action> = None;

        for r in 0..h {
            for c in 0..w {
                let idx = (r * w + c) as usize;
                if obs.owner_grid[idx] != 1 { continue; }
                let src_army = obs.army_grid[idx];
                if src_army <= 1 { continue; }

                for d in 0..4 {
                    let nr = r + DR[d];
                    let nc = c + DC[d];
                    if nr < 0 || nr >= h || nc < 0 || nc >= w { continue; }
                    let didx = (nr * w + nc) as usize;
                    let dtype = obs.type_grid[didx];
                    if dtype == 2 || dtype == 5 { continue; }   // impassable

                    let m = Action { pass: 0, row: r, col: c, dir: d as i32, split: 0 };
                    if first_valid.is_none() {
                        first_valid = Some(m);
                    }

                    let dest_owner = obs.owner_grid[didx];
                    let dest_army = obs.army_grid[didx];
                    if src_army <= dest_army + 1 { continue; }

                    let is_opp = dest_owner == 2;
                    let is_visible_neutral = dest_owner == 0 && dtype != 0 && dtype != 5;
                    let is_expansion = is_opp || is_visible_neutral;

                    let mut score = src_army as f64;
                    if is_expansion { score *= 10.0; }
                    if is_opp       { score *= 2.0; }

                    if score > best_score {
                        best_score = score;
                        best = m;
                    }
                }
            }
        }

        if best_score > 0.0 { best }
        else if let Some(fv) = first_valid { fv }
        else { PASS }
    }
}
